### PR TITLE
[mono] iOS Test runner

### PR DESCRIPTION
**Pull request by EgorBo**
It's currently WIP, it doesn't use runtime packs yet but it's already possible to run any test suite (well, any without remote-executor based tests inside, see https://github.com/dotnet/arcade/issues/5129 such tests usually have `<IncludeRemoteExecutor>true</IncludeRemoteExecutor>` inside csprojs). It uses dotnet/xharness test runner, copies everything including the whole BCL into a test folder and runs the AOT/Bundle generation.
  
Steps to run a test suite:
```sh
# make sure everything is built for iOS (can be arm64 instead of x64 for devices and fullAOT)
./build.sh -c Release -os iOS -arch x64 -subset Mono+Libs+Libs.Tests /p:DisableCrossgen=true

# TestName can be any test suite, e.g. System.Collections.Tests
./dotnet.sh msbuild src/mono/mono.proj /t:BuildAppleTestBundle \
/p:Configuration=Release /p:TestName=System.Buffers.Tests \
/p:TargetOS=iOS /p:TargetArchitecture=x64
```
for x64 it should build everything and launch an iphone simulator with tests. For arm64 devices it needs `xharness` tool (with `mlaunch` inside - it's not yet there) and a provisioning profile.
If you want to try arm64 now you can run the script above ^ (replace x64 with arm64) and open the xcode project (generated) and deploy to a personal device with auto-generated provisioning. Some tests might crash on arm64, e.g.:
```
Assertion at /Users/egorbo/prj/runtime-3/src/mono/mono/metadata/class.c:1845, 
condition `method->is_inflated' not met
```
or some missing *aotdata stuff. Arm64 needs some work.

image

I guess the expected workflow is the following:
```
cd src/libraries/System.Buffers/tests
dotnet publish -c Release -r ios-x64
dotnet tool xharness test ios --app path/to/app/bunndle.app
```

### CODE REVIEW

**Comment by safern** [comment]
Adding my 2 cents here. I think we should unify this with our current testing infrastructure/targets to run libraries tests instead of diverging paths by having some stuff under mono and some other under libraries. 

We could use a lot of what we have by adding some targets/commands to execute the app bundler and invoking the tests and what not.

**Comment by safern** [comment]
Adding my 2 cents here. I think we should unify this with our current testing infrastructure/targets to run libraries tests instead of diverging paths by having some stuff under mono and some other under libraries. 

We could use a lot of what we have by adding some targets/commands to execute the app bundler and invoking the tests and what not.

**Comment by steveisok** [comment]
@safern At the very least, I would expect this PR to drive towards that.

**Comment by safern** [comment]
> @safern At the very least, I would expect this PR to drive towards that.

Yes, I think it should as a first step.

**Comment by marek-safar** [comment]
>  I think we should unify this with our current testing infrastructure/targets to run libraries tests instead of diverging paths by having some stuff under mono and some other under libraries.

@safern what would you like to be moved and to where? For now, only very specific libraries configurations are supported but the same infrastructure will be used for all other tests in dotnet/runtime so not sure if it makes sense to move it libraries as it's not libraries specific.

File: src/mono/msbuild/AppleAppBuilder/Xcode.cs
Comment (marek-safar) [inline src/mono/msbuild/AppleAppBuilder/Xcode.cs:92]: I think should be explicit to catch newly added libs which could break XI

File: src/mono/mono.proj
Comment (marek-safar) [inline src/mono/mono.proj:921]: I don't think this should be done here

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.csproj
Comment (marek-safar) [inline src/mono/msbuild/AppleTestRunner/AppleTestRunner.csproj:9]: Should go to eng/Versions.props or something like that

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs
Comment (marek-safar) [inline src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:49]: Why is not passes as argument from $(TestName) ?

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.csproj
Comment (akoeplinger) [inline src/mono/msbuild/AppleTestRunner/AppleTestRunner.csproj:6]: why DebugType full? should be portable which is the default.

**Comment by safern** [comment]
> @safern what would you like to be moved and to where? For now, only very specific libraries configurations are supported but the same infrastructure will be used for all other tests in dotnet/runtime so not sure if it makes sense to move it libraries as it's not libraries specific

I’d like to have this hooked up under /eng/testing and have the targets and all underlying logic there. Even though it will not be used by all configurations what we’re running are libraries tests. I don’t like all this manual copying of files and duplicating path calculation and all things involved here. We could hook it up in the infra we already have to run tests and everything would be much simpler. Instead of having to run specific mono targets I envision this as `dotnet build /t:test /p:TargetOS=iOS /p:TargetArchitecture=x64` on a test project and the test project will be executed on the simulator.

Or `build libs.tests -test -os iOS -arch x64` and all the tests will be executed on the simulator. I want to be consistent on how we run tests for all configurations this makes dev experience easier rather than having to remember different commands and what project to call etc. 

I think this is fine as is for the purpose of this PR, but we should eventually hook it up as o suggest above. I can definitely help with that, I started taking a look at how all this works yesterday and I think it is possible.

**Comment by marek-safar** [comment]
@steveisok could you ensure @safern concerns are addressed

**Comment by safern** [comment]
cc: @ViktorHofer @danmosemsft

**Comment by steveisok** [comment]
@safern After talking w/ Egor & Alex, I'd like to land this PR as is so that we have a runner to use.  Then in a separate PR, we can hook into what you outlined.

**Comment by safern** [comment]
> @safern After talking w/ Egor & Alex, I'd like to land this PR so that we have a runner to use. Then in a separate PR, we can hook into what you outlined.

Yeah I don't want to block this PR. I just wanted to raise my concerns as a heads up that we should envision hooking it up the same way we already run libraries tests and of course happy to help with doing that.

**Review by akoeplinger — APPROVED** [review]

File: src/mono/mono.proj
Comment (danmosemsft) [inline src/mono/mono.proj:926]: `SkipUnchangedFiles="true"` for all these? Don't know why that 's not the default.

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs
Comment (danmosemsft) [inline src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:63]:
```suggestion
            Console.WriteLine("Test libs were not found (*.Tests.dll was not found in {Environment.CurrentDirectory})");
```

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs
Comment (danmosemsft) [inline src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:33]: `iOSApplicatonEntryPoint` -- is this a typo? `iOSApplicationEntryPoint`

File: src/mono/msbuild/AppleAppBuilder/AppleAppBuilder.cs
Comment (danmosemsft) [inline src/mono/msbuild/AppleAppBuilder/AppleAppBuilder.cs:37]: What does ` ""!` achieve here? Is this the "null-forgiving" operator?

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs
Comment (danmosemsft) [reply src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:33]: ah, it's in dotnet/xharness. I didn't know about this repo 😸

Comment (EgorBo) [reply src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:33]: cc @mandel-macaque ^

File: src/mono/msbuild/AppleAppBuilder/AppleAppBuilder.cs
Comment (EgorBo) [reply src/mono/msbuild/AppleAppBuilder/AppleAppBuilder.cs:37]: These properties are guaranteed to be non-null always (due to `[Required]` attribute) the compiler forces me to do inline init to leave them non-nullable. I wonder how @stephentoub solves such cases.

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs
Comment (EgorBo) [reply src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:33]: also, while you are there could you please add events for "test is started", "test is finished (with result)" ?

File: src/mono/msbuild/AppleAppBuilder/Templates/main-console.m
Comment (EgorBo) [inline src/mono/msbuild/AppleAppBuilder/Templates/main-console.m:87]: Some hack to automatically scroll to the bottom in UITextView

Comment (akoeplinger) [reply src/mono/msbuild/AppleAppBuilder/Templates/main-console.m:87]: why do you need to toggle setScrollEnabled? is the scrollRectToVisible not enough?

Comment (EgorBo) [reply src/mono/msbuild/AppleAppBuilder/Templates/main-console.m:87]: @akoeplinger I don't know but without it doesn't work properly, I tried different hacks from SO this one works as expected (https://stackoverflow.com/questions/19124037/scroll-to-bottom-of-uitextview-erratic-in-ios-7/20989956#20989956)

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs
Comment (mandel-macaque) [reply src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:14]: @danmosemsft I'm known for my typos, I'll get that fixed! @EgorBo  https://github.com/dotnet/xharness/issues/36

File: src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs
Comment (safern) [inline src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:65]: Why is the null forgiving operator needed here?

Comment (EgorBo) [reply src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:65]: @safern I forgot to remove it

Comment (EgorBo) [reply src/mono/msbuild/AppleTestRunner/AppleTestRunner.cs:65]: Will fix it in the next PR

**Comment by akoeplinger** [comment]
`runtime (Installer Build and Test Linux_x64 Debug)` passed on AzDO, the status didn't make it to GitHub: https://dev.azure.com/dnceng/public/_build/results?buildId=603747&view=logs&j=0d59b9e3-002e-58b9-0bdb-d3cd1e08882b&t=90f7f7ef-163c-557b-3338-db1b69e0bef4&l=21

**Comment by EgorBo** [comment]
`runtime (Installer Build and Test Linux_x64 Debug)` passed on AzDO but the status didn't make it to github UI, so everything is green. Going to merge it and improve as we go.

**Pull request merged**
